docs/set.rst

```diff
@@ -0,0 +1,471 @@
+Sets
+====
+
+:haddock:`unordered-containers`
+
+:haddock:`unordered-containers/Data.HashSet`
+
+:haddock:`unordered-containers/Data.HashSet#empty`
+
+.. highlight:: haskell
+
+Sets allow you to store *unique*, *ordered* elements, providing efficient
+insertion, lookups, deletions, and set operations. There are two implementations
+provided by the ``containers`` package: :haddock:`containers/Data.Set` and
+:haddock:`containers/Data.IntSet`. Use ``IntSet`` if you are storing,
+well... ``Int`` s; both of these implementations are *immutable*.
+
+::
+
+    data Set element = ...
+
+    data IntSet = ...
+
+.. IMPORTANT::
+   ``Set`` relies on the `element` type having instances of the ``Eq`` and
+   ``Ord`` typeclass for its internal representation. These are already defined
+   for builtin types, and if you are using your own data type you can use the
+   `deriving
+   <https://en.wikibooks.org/wiki/Haskell/Classes_and_types#Deriving>`_
+   mechanism.
+
+
+Short Example
+-------------
+
+The following GHCi session shows some of the basic set functionality::
+
+    import qualified Data.Set as Set
+
+    let dataStructures = Set.fromList ["Set", "Map", "Graph", "Sequence"]
+
+    -- Check if "Map" and "Trie" are in the set of data structures.
+    Set.member "Map" dataStructures
+    > True
+
+    Set.member "Trie" dataStructures
+    > False
+
+
+    -- Add "Trie" to our original set of data structures.
+    let moreDataStructures = Set.insert "Trie" dataStructures
+
+    Set.member "Trie" moreDataStructures
+    > True
+
+
+    -- Remove "Graph" froum our original set of data structures.
+    let fewerDataStructures = Set.delete "Graph" dataStructures
+
+    Set.toAscList fewerDataStructures
+    > ["Map","Sequence","Set"]
+
+
+    -- Create a new set and combine it with our original set.
+    let unorderedDataStructures = Set.fromList ["HashSet", "HashMap"]
+
+    Set.union dataStructures unorderedDataStructures
+    > fromList ["Graph","HashMap","HashSet","Map","Sequence","Set"]
+
+
+.. TIP:: You can use the `OverloadedLists
+	 <https://ghc.haskell.org/trac/ghc/wiki/OverloadedLists>`_ extension so
+	 you don't need to write ``fromList [1, 2, 3]`` everywhere. Instead you
+	 can just write ``[1, 2, 3]`` and if the function is expecting a set it
+	 will be converted automatically! The code here will continue to use
+	 ``fromList`` for clarity though.
+
+
+Importing Set and IntSet
+------------------------
+
+When using ``Set`` or ``IntSet`` in a Haskell source file you should always use
+a ``qualified`` import because these modules export names that clash with the
+standard Prelude (you can import the type constructor on its own though!):
```

Comment:
Also here, I would expand from just talking about the type constructor to any additional things you desire.

Reply:
Oops, these were here when I was testing the custom role and I forgot to remove them.

Reply:
Done.